File: tv_app_native/app/src/main/java/ir/talalive/tv/BoardModel.kt
package ir.talalive.tv

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject

data class PriceRow(
    val symbol: String,
    val title: String,
    val price: String,
    val unit: String,
    val changeDirection: Int,   // -1 (down), 0 (flat), +1 (up)
    val changeText: String?,
    val isStale: Boolean
)

data class ProductItem(
    val id: String,
    val title: String,
    val weightGram: String?,
    val finalPrice: String?,
    val laborFee: String?,
    val imageUrls: List<String>
)

data class BoardModel(
    val username: String,
    val shopName: String,
    val phone: String,
    val themeMode: String,
    val sliderIntervalSec: Int,
    val rows: List<PriceRow>,
    val products: List<ProductItem>,
    val updatedAtText: String,
    val dataAgeSeconds: Int,
    val isStale: Boolean,
    val refreshIntervalSeconds: Int,
    val rawJson: String
) {
    companion object {
        private const val TAG = "TalaTV.BoardModel"

        fun parse(raw: String): BoardModel? {
            if (raw.isBlank()) return null

            return try {
                val root = JSONObject(raw)

                val username = root.optString("username", "")
                val updatedAt = root.optString("updatedAt", "")
                val dataAgeSeconds = root.optInt("dataAgeSeconds", 0)
                val isStale = root.optBoolean("isStale", false)
                val refreshSec = root.optInt("refreshIntervalSeconds", 60).coerceIn(5, 300)

                val settingsObj = root.optJSONObject("settings")
                val shopName = settingsObj?.optString("shop_name", "گالری طلا")?.ifBlank { "گالری طلا" } ?: "گالری طلا"
                val phone = settingsObj?.optString("phone", "") ?: ""
                val themeMode = settingsObj?.optString("theme_mode", "light-modern") ?: "light-modern"
                val sliderInterval = settingsObj?.optInt("slider_interval_sec", 8)?.coerceIn(3, 60) ?: 8

                // 1. Parse priceFeed map: symbol -> PriceFeedItem
                val priceFeedMap = LinkedHashMap<String, RawPrice>()
                val feedArr = root.optJSONArray("priceFeed") ?: JSONArray()
                for (i in 0 until feedArr.length()) {
                    val pObj = feedArr.optJSONObject(i) ?: continue
                    val symbol = pObj.optString("symbol", "").trim()
                    if (symbol.isEmpty()) continue

                    val label = pObj.optString("label", symbol)
                    val rawVal = pObj.opt("value")
                    val valueStr = cleanNumericValue(rawVal)
                    val unit = pObj.optString("unit", "تومان")
                    val dirStr = pObj.optString("direction", "flat").lowercase()
                    val dirInt = when (dirStr) {
                        "up" -> 1
                        "down" -> -1
                        else -> 0
                    }
                    val chgPct = pObj.optString("change_percent", "").trim()
                    val chgText = if (chgPct.isNotEmpty() && chgPct != "0" && chgPct != "0.00") "$chgPct٪" else null
                    val itemStale = pObj.optBoolean("is_stale", false)

                    priceFeedMap[symbol] = RawPrice(symbol, label, valueStr, unit, dirInt, chgText, itemStale)
                }

                // 2. Parse displayItems if present and join with priceFeed
                val rows = ArrayList<PriceRow>()
                val displayItemsArr = root.optJSONArray("displayItems")

                if (displayItemsArr != null && displayItemsArr.length() > 0) {
                    val itemsList = ArrayList<DisplayItemEntry>()
                    for (i in 0 until displayItemsArr.length()) {
                        val dObj = displayItemsArr.optJSONObject(i) ?: continue
                        val key = dObj.optString("key", "").trim()
                        if (key.isEmpty()) continue
                        val enabled = dObj.optBoolean("enabled", true)
                        if (!enabled) continue
                        val label = dObj.optString("label", "").trim()
                        val order = dObj.optInt("order", i)
                        itemsList.add(DisplayItemEntry(key, label, order))
                    }
                    itemsList.sortBy { it.order }

                    for (entry in itemsList) {
                        val price = priceFeedMap[entry.key]
                        if (price != null) {
                            val title = entry.customLabel.ifBlank { price.defaultLabel }
                            rows.add(
                                PriceRow(
                                    symbol = price.symbol,
                                    title = title,
                                    price = price.value,
                                    unit = price.unit,
                                    changeDirection = price.changeDirection,
                                    changeText = price.changeText,
                                    isStale = price.isStale
                                )
                            )
                        }
                    }
                } else {
                    // Fallback: use priceFeed directly in order
                    for ((_, price) in priceFeedMap) {
                        rows.add(
                            PriceRow(
                                symbol = price.symbol,
                                title = price.defaultLabel,
                                price = price.value,
                                unit = price.unit,
                                changeDirection = price.changeDirection,
                                changeText = price.changeText,
                                isStale = price.isStale
                            )
                        )
                    }
                }

                // 3. Parse products
                val products = ArrayList<ProductItem>()
                val prodArr = root.optJSONArray("products")
                if (prodArr != null) {
                    for (i in 0 until prodArr.length()) {
                        val pObj = prodArr.optJSONObject(i) ?: continue
                        val pId = pObj.optString("id", "")
                        val pTitle = pObj.optString("title", "")
                        val isVisible = pObj.opt("is_visible")
                        val visible = isVisible == true || isVisible == "1" || isVisible == 1
                        if (!visible) continue

                        val imgList = ArrayList<String>()
                        val imgArr = pObj.optJSONArray("images")
                        if (imgArr != null) {
                            for (j in 0 until imgArr.length()) {
                                val iObj = imgArr.optJSONObject(j) ?: continue
                                val u = iObj.optString("url", "").trim()
                                if (u.isNotEmpty()) imgList.add(u)
                            }
                        }
                        if (imgList.isNotEmpty()) {
                            val pWeight = pObj.optString("weight_gram", "").trim().ifEmpty { null }
                            val pFinal = cleanNumericValue(pObj.opt("final_price")).let { if (it == "0") null else it }
                            val pLabor = pObj.optString("labor_fee", "").trim().ifEmpty { null }
                            products.add(ProductItem(pId, pTitle, pWeight, pFinal, pLabor, imgList))
                        }
                    }
                }

                BoardModel(
                    username = username,
                    shopName = shopName,
                    phone = phone,
                    themeMode = themeMode,
                    sliderIntervalSec = sliderInterval,
                    rows = rows,
                    products = products,
                    updatedAtText = updatedAt,
                    dataAgeSeconds = dataAgeSeconds,
                    isStale = isStale,
                    refreshIntervalSeconds = refreshSec,
                    rawJson = raw
                )
            } catch (t: Throwable) {
                Log.e(TAG, "Failed to parse BoardModel", t)
                null
            }
        }

        private fun cleanNumericValue(raw: Any?): String {
            if (raw == null) return "0"
            val s = raw.toString().trim()
            if (s.isEmpty()) return "0"

            // If it ends with .00 or .0, drop the trailing zeros for clean integer display
            return if (s.contains(".")) {
                val parts = s.split(".")
                if (parts.size == 2 && (parts[1] == "0" || parts[1] == "00")) {
                    parts[0]
                } else {
                    s
                }
            } else {
                s
            }
        }
    }

    private data class RawPrice(
        val symbol: String,
        val defaultLabel: String,
        val value: String,
        val unit: String,
        val changeDirection: Int,
        val changeText: String?,
        val isStale: Boolean
    )

    private data class DisplayItemEntry(
        val key: String,
        val customLabel: String,
        val order: Int
    )
}

File: tv_app_flutter/lib/models/board_model.dart
import '../utils/persian_utils.dart';

class PriceRow {
  final String symbol;
  final String title;
  final String sellPrice;
  final String? buyPrice;
  final String unit;
  final int changeDirection; // -1 (down), 0 (flat), +1 (up)
  final String? changePercent;
  final String? changeValue;
  final bool isStale;

  const PriceRow({
    required this.symbol,
    required this.title,
    required this.sellPrice,
    this.buyPrice,
    required this.unit,
    required this.changeDirection,
    this.changePercent,
    this.changeValue,
    required this.isStale,
  });
}

class ProductItem {
  final String id;
  final String title;
  final String? weightGram;
  final String? finalPrice;
  final String? laborFee;
  final String? profitValue;
  final String? profitType;
  final String? badge;
  final int? sortOrder;
  final List<String> imageUrls;

  const ProductItem({
    required this.id,
    required this.title,
    this.weightGram,
    this.finalPrice,
    this.laborFee,
    this.profitValue,
    this.profitType,
    this.badge,
    this.sortOrder,
    required this.imageUrls,
  });

  String getDisplayPrice(num gold18Price) {
    final w = num.tryParse(weightGram ?? '') ?? 0;
    final l = num.tryParse(laborFee ?? '') ?? 0;
    final pv = num.tryParse(profitValue ?? '') ?? 0;
    if (gold18Price > 0 && w > 0) {
      final base = (gold18Price * w) + l;
      num profit = 0;
      if (profitType == 'percent') {
        profit = base * (pv / 100);
      } else if (profitType == 'amount_per_gram') {
        profit = pv * w;
      } else {
        profit = pv;
      }
      final calc = (base + profit).round();
      if (calc > 0) return calc.toString();
    }
    if (finalPrice != null && finalPrice!.isNotEmpty && finalPrice != '0') {
      return finalPrice!;
    }
    return '۰';
  }
}

int _parseInt(dynamic val, int def) {
  if (val == null) return def;
  if (val is num) return val.toInt();
  if (val is String) {
    return int.tryParse(val) ?? (double.tryParse(val)?.toInt() ?? def);
  }
  return def;
}

bool _parseBool(dynamic val, bool def) {
  if (val == null) return def;
  if (val is bool) return val;
  if (val is num) return val != 0;
  if (val is String) {
    final s = val.toLowerCase().trim();
    if (s == 'true' || s == '1') return true;
    if (s == 'false' || s == '0') return false;
  }
  return def;
}

class BoardModel {
  final String username;
  final String shopName;
  final String subtitle;
  final String phone;
  final String? instagram;
  final String? rubika;
  final String? qrLink;
  final String? qrLabel;
  final String? qrDesc;
  final String? cityFullDisplay;
  final String? cityName;
  final String? galleryDisplayName;
  final String? bingWallpaperUrl;
  final String themeMode;
  final int sliderIntervalSec;
  final String customMessage;
  final List<PriceRow> rows;
  final List<ProductItem> products;
  final String updatedAtText;
  final int dataAgeSeconds;
  final bool isStale;
  final int refreshIntervalSeconds;
  final String rawJson;

  const BoardModel({
    required this.username,
    required this.shopName,
    required this.subtitle,
    required this.phone,
    this.instagram,
    this.rubika,
    this.qrLink,
    this.qrLabel,
    this.qrDesc,
    this.cityFullDisplay,
    this.cityName,
    this.galleryDisplayName,
    this.bingWallpaperUrl,
    required this.themeMode,
    required this.sliderIntervalSec,
    required this.customMessage,
    required this.rows,
    required this.products,
    required this.updatedAtText,
    required this.dataAgeSeconds,
    required this.isStale,
    required this.refreshIntervalSeconds,
    required this.rawJson,
  });

  factory BoardModel.fromJson(Map<String, dynamic> json, String rawString) {
    final username = json['username']?.toString() ?? '';
    final updatedAt = json['updatedAt']?.toString() ?? '';
    final dataAgeSec = _parseInt(json['dataAgeSeconds'], 0);
    final isStale = _parseBool(json['isStale'], false);
    final refreshSec = _parseInt(json['refreshIntervalSeconds'], 60).clamp(5, 300);

    final rawSettings = json['settings'];
    final settings = rawSettings is Map ? Map<String, dynamic>.from(rawSettings) : null;
    final shopName = json['shopName']?.toString().trim().isNotEmpty == true
        ? json['shopName'].toString().trim()
        : (settings?['shop_name']?.toString().trim().isNotEmpty == true
            ? settings!['shop_name'].toString().trim()
            : 'طلا و جواهر طلالایو');
    final galleryDisplayName = json['galleryDisplayName']?.toString().trim().isNotEmpty == true
        ? json['galleryDisplayName'].toString().trim()
        : shopName;

    String? bingWallpaperUrl;
    final rawBing = json['bingWallpaper'];
    if (rawBing is Map) {
      final url = rawBing['url']?.toString();
      if (url != null && url.isNotEmpty) {
        bingWallpaperUrl = url.startsWith('/') ? 'https://talalive.ir$url' : url;
      }
    } else if (rawBing is String && rawBing.isNotEmpty) {
      bingWallpaperUrl = rawBing.startsWith('/') ? 'https://talalive.ir$rawBing' : rawBing;
    }
    final subtitle = settings?['subtitle']?.toString() ?? 'تابلوی رسمی نرخ لحظه‌ای طلا، سکه و ارز';
    final phone = settings?['phone']?.toString() ?? '';
    final instagram = settings?['instagram']?.toString();
    final rubika = settings?['rubika']?.toString();
    final qrLink = settings?['qr_link']?.toString();
    final qrLabel = settings?['qr_label']?.toString();
    final qrDesc = settings?['qr_desc']?.toString();
    final rawCitySlug = json['citySlug']?.toString() ?? settings?['city_slug']?.toString();
    final rawCityName = json['cityName']?.toString() ?? settings?['city_name']?.toString();
    final rawCityFull = json['cityFullDisplay']?.toString() ?? settings?['city_full_display']?.toString();

    final cityFullDisplay = (rawCityFull != null && rawCityFull.trim().isNotEmpty && rawCityFull.trim() != 'null')
        ? PersianUtils.resolveCityName(rawCityFull.trim())
        : ((rawCityName != null && rawCityName.trim().isNotEmpty && rawCityName.trim() != 'null')
            ? PersianUtils.resolveCityName(rawCityName.trim())
            : PersianUtils.resolveCityName(rawCitySlug));
    final cityName = (rawCityName != null && rawCityName.trim().isNotEmpty && rawCityName.trim() != 'null')
        ? PersianUtils.resolveCityName(rawCityName.trim())
        : PersianUtils.resolveCityName(rawCitySlug);
    final themeMode = settings?['theme_mode']?.toString() ?? 'luxury-dark';
    final sliderInterval = _parseInt(settings?['slider_interval_sec'], 8).clamp(3, 60);
    final customMessage = settings?['custom_message']?.toString() ??
        'به سامانه تابلوی هوشمند نرخ لحظه‌ای طلالایو خوش آمدید • نمایش دقیق و لحظه‌ای مظنه طلا، سکه و مسکوکات';

    // 1. PriceFeed map
    final priceFeedMap = <String, Map<String, dynamic>>{};
    final feedList = json['priceFeed'] as List<dynamic>? ?? [];
    for (final item in feedList) {
      if (item is Map) {
        final map = Map<String, dynamic>.from(item);
        final symbol = map['symbol']?.toString().trim() ?? '';
        if (symbol.isNotEmpty) {
          priceFeedMap[symbol] = map;
        }
      }
    }

    // 2. Parse DisplayItems or fallback to priceFeed
    final rows = <PriceRow>[];
    final displayItems = json['displayItems'] as List<dynamic>?;

    if (displayItems != null && displayItems.isNotEmpty) {
      final entries = <Map<String, dynamic>>[];
      for (var i = 0; i < displayItems.length; i++) {
        final d = displayItems[i];
        if (d is Map) {
          final dMap = Map<String, dynamic>.from(d);
          final enabled = dMap['enabled'] != false;
          if (enabled) {
            entries.add(dMap);
          }
        }
      }

      for (final entry in entries) {
        final key = entry['key']?.toString().trim() ?? '';
        final feed = priceFeedMap[key];
        if (feed != null) {
          final customLabel = entry['label']?.toString().trim() ?? '';
          final defaultLabel = feed['label']?.toString().trim() ?? key;
          final title = customLabel.isNotEmpty ? customLabel : defaultLabel;
          final val = feed['value']?.toString() ?? '۰';
          final buyVal = feed['buy_price']?.toString();
          final unit = feed['unit']?.toString() ?? 'تومان';
          final dirStr = feed['direction']?.toString().toLowerCase() ?? 'flat';
          final dirInt = dirStr == 'up' ? 1 : (dirStr == 'down' ? -1 : 0);
          final chgPct = feed['change_percent']?.toString().trim();
          final chgVal = feed['change_value']?.toString().trim();
          final itemStale = _parseBool(feed['is_stale'], false);

          rows.add(PriceRow(
            symbol: key,
            title: title,
            sellPrice: val,
            buyPrice: buyVal,
            unit: unit,
            changeDirection: dirInt,
            changePercent: chgPct,
            changeValue: chgVal,
            isStale: itemStale,
          ));
        }
      }
    } else {
      // Fallback: Direct priceFeed
      for (final entry in priceFeedMap.entries) {
        final feed = entry.value;
        final title = feed['label']?.toString().trim() ?? entry.key;
        final val = feed['value']?.toString() ?? '۰';
        final buyVal = feed['buy_price']?.toString();
        final unit = feed['unit']?.toString() ?? 'تومان';
        final dirStr = feed['direction']?.toString().toLowerCase() ?? 'flat';
        final dirInt = dirStr == 'up' ? 1 : (dirStr == 'down' ? -1 : 0);
        final chgPct = feed['change_percent']?.toString().trim();
        final chgVal = feed['change_value']?.toString().trim();
        final itemStale = _parseBool(feed['is_stale'], false);

        rows.add(PriceRow(
          symbol: entry.key,
          title: title,
          sellPrice: val,
          buyPrice: buyVal,
          unit: unit,
          changeDirection: dirInt,
          changePercent: chgPct,
          changeValue: chgVal,
          isStale: itemStale,
        ));
      }
    }

    // 3. Products
    final products = <ProductItem>[];
    final prodList = json['products'] as List<dynamic>?;
    if (prodList != null) {
      for (final p in prodList) {
        if (p is Map) {
          final pMap = Map<String, dynamic>.from(p);
          final visible = _parseBool(pMap['is_visible'], true);
          if (!visible) continue;

          final id = pMap['id']?.toString() ?? '';
          final title = pMap['title']?.toString() ?? '';
          final weight = pMap['weight_gram']?.toString();
          final price = pMap['final_price']?.toString();
          final labor = pMap['labor_fee']?.toString();
          final profitVal = pMap['profit_value']?.toString();
          final profitTyp = pMap['profit_type']?.toString();
          final badge = pMap['badge']?.toString();
          final sortOrder = _parseInt(pMap['sort_order'], 0);

          final imgList = <String>[];
          // 1. Eloquent 'images' relationship: [{url: "..."}, ...]
          final imagesList = pMap['images'] as List<dynamic>?;
          if (imagesList != null) {
            for (final img in imagesList) {
              if (img is Map) {
                final u = img['url']?.toString().trim();
                if (u != null && u.isNotEmpty) {
                  imgList.add(u);
                }
              } else if (img is String && img.trim().isNotEmpty) {
                imgList.add(img.trim());
              }
            }
          }
          // 2. Fallback: image_urls
          final rawImgs = pMap['image_urls'] as List<dynamic>?;
          if (rawImgs != null) {
            for (final u in rawImgs) {
              if (u != null && u.toString().trim().isNotEmpty) {
                imgList.add(u.toString().trim());
              }
            }
          }
          // 3. Fallback: image_url
          if (imgList.isEmpty) {
            final single = pMap['image_url']?.toString().trim();
            if (single != null && single.isNotEmpty) {
              imgList.add(single);
            }
          }

          products.add(ProductItem(
            id: id,
            title: title,
            weightGram: weight,
            finalPrice: price,
            laborFee: labor,
            profitValue: profitVal,
            profitType: profitTyp,
            badge: badge,
            sortOrder: sortOrder,
            imageUrls: imgList,
          ));
        }
      }
    }

    return BoardModel(
      username: username,
      shopName: shopName,
      subtitle: subtitle,
      phone: phone,
      instagram: instagram,
      rubika: rubika,
      qrLink: qrLink,
      qrLabel: qrLabel,
      qrDesc: qrDesc,
      cityFullDisplay: cityFullDisplay,
      cityName: cityName,
      galleryDisplayName: galleryDisplayName,
      bingWallpaperUrl: bingWallpaperUrl,
      themeMode: themeMode,
      sliderIntervalSec: sliderInterval,
      customMessage: customMessage,
      rows: rows,
      products: products,
      updatedAtText: updatedAt,
      dataAgeSeconds: dataAgeSec,
      isStale: isStale,
      refreshIntervalSeconds: refreshSec,
      rawJson: rawString,
    );
  }
}
